Improve translation loading

Holzhaus opened the pull request:
While using developer mode, I saw some suspicious log messages regarding failed translations when using the `en` locale. When looking at the code I realized that mixxx doesn't even attempt to load them (which makes sense) but the error message is misleading.

I cleaned up the translation handling a bit, added some error messages and made them more meaningful. I based this on 2.3 but If you want to postpone this for 2.4 we can do that as well.

Holzhaus committed:
Translations: Print warning if translations failed to load

Holzhaus committed:
Translations: Improve locale override logic

Holzhaus committed:
Translations: Use helper method for unified error handling

daschuer commented on src/util/translations.h:
```diff
@@ -13,92 +13,101 @@ namespace mixxx {
-            QLocale::setDefault(QLocale(systemLocale));
-        } else {
-            QLocale::setDefault(QLocale(userLocale));
+        QLocale locale = QLocale();
```
This assignment is redundant.
Please add a comment that we investigate the default value.

Holzhaus replied:
> Please add a comment that we investigate the default value.

What do you mean?

From the [Qt docs](https://doc.qt.io/qt-5/qlocale.html#QLocale):
> Constructs a QLocale object initialized with the default locale. If no default locale was set using setDefault(), this locale will be the same as the one returned by system().

Holzhaus replied:
I added that as a comment.

daschuer reviewed (changes requested):
2.3 is fine for me. I have left some comments.

Holzhaus commented on src/util/translations.h:
```diff
@@ -13,92 +13,101 @@ namespace mixxx {
+        // Set the default locale, if specified via command line or config
+        // file. If neither is the case, the default locale will be the
+        // system's locale.
+        {
```
It's not really necessary but `customLocaleString` will be dropped after it. I can just remove it if it's confusing.

Holzhaus committed:
Translations: Simplify code (remove scope and redundant default constrs)

Holzhaus committed:
Translations: Use default value when reading from config

Holzhaus commented on src/util/translations.h:
```diff
@@ -13,92 +13,99 @@ namespace mixxx {
+                QStringLiteral("qt"),
+                QLibraryInfo::location(QLibraryInfo::TranslationsPath));
 
         // Load Qt translations for this locale from the Mixxx translations
         // folder.
```
Is this even relevant anymore? I don't think there exist...

```
$ ls -l ../res/translations/ | grep qt | wc -l
0
```

Holzhaus replied:
Wait, is this only relevant on Windows?

Holzhaus replied:
For these translations, I changed to loglevel to Debug instead of warning.

Holzhaus committed:
Translations: Use Debug instead of Warning loglevel for mixxx qt strings

daschuer commented:
Warning [Main]: Unsupported locale "ceb"

When I use "System" I get confusing messages on Ubuntu Focal: 

```
Debug [Main]: Found and will use default keyboard preset "/home/daniel/workspace/2.3/res/keyboard/en_US.kbd.cfg"
Debug [Main]: Loading resources from  "/home/daniel/workspace/2.3/res/"
Debug [Main]: Loaded "qt" translations for locale "de_DE" from "/usr/share/qt5/translations"
Debug [Main]: Failed to load "qt" translations for locale "de_DE" from "/home/daniel/workspace/2.3/res/translations/"
Debug [Main]: Loaded "mixxx" translations for locale "de_DE" from "/home/daniel/workspace/2.3/res/translations/"
```

Holzhaus commented:
> Warning [Main]: Unsupported locale "ceb"

I cannot reproduce this. I guess your config was messed up and contained an invalid string for the locale setting.

> When I use "System" I get confusing messages on Ubuntu Focal:
> 
> ```
> Debug [Main]: Found and will use default keyboard preset "/home/daniel/workspace/2.3/res/keyboard/en_US.kbd.cfg"
> Debug [Main]: Loading resources from  "/home/daniel/workspace/2.3/res/"
> Debug [Main]: Loaded "qt" translations for locale "de_DE" from "/usr/share/qt5/translations"
> Debug [Main]: Failed to load "qt" translations for locale "de_DE" from "/home/daniel/workspace/2.3/res/translations/"
> Debug [Main]: Loaded "mixxx" translations for locale "de_DE" from "/home/daniel/workspace/2.3/res/translations/"
> ```

What is confusing about it? This looks correct to me.

It's normal that loading `qt` translations from the mixxx translation directory fails, i think that is only used on windows (an macOS maybe?). See https://github.com/mixxxdj/mixxx/pull/3826/commits/e847f58f5a268b104206a271ab709ae5ff7981c1.

esbrandt commented:
> Warning [Main]: Unsupported locale "ceb"

Delete 
https://github.com/mixxxdj/mixxx/blob/2.3/res/translations/mixxx_ceb.qm
https://github.com/mixxxdj/mixxx/blob/2.3/res/translations/mixxx_ceb.ts
Both files have never been touched since ages, and the language is not even supported on Transifex.

Holzhaus committed:
res/translations: Delete outdates/unused "ceb" translations

See https://github.com/mixxxdj/mixxx/pull/3826#issuecomment-834334745.

Holzhaus commented:
> > Warning [Main]: Unsupported locale "ceb"
> 
> Delete
> https://github.com/mixxxdj/mixxx/blob/2.3/res/translations/mixxx_ceb.qm
> https://github.com/mixxxdj/mixxx/blob/2.3/res/translations/mixxx_ceb.ts
> Both files have never been touched since ages, and the language is not even supported on Transifex.

Done.

Holzhaus commented:
Anything else to do here?

esbrandt commented:
LGTM

Holzhaus commented:
ping @daschuer

uklotzde reviewed (approved):
Very helpful comments. LGTM

@daschuer Ping

Holzhaus commented:
> Why do we tray to load the qt translations from the res path when we already have loaded them from the system? Can we make the local lookup conditionally? Or load directly from the correct location target depending. 
> 
> On windows we install also qtbase and qthelp translations among others in the translations folder but the corresponding install statement is missing. I am afraid they are not loaded, are they? Are they loaded under Windows. 

I didn't change this. If it worked before, I see no reason why it should not work now.

We could check if the system install translations call succeeded and skip the other one, but I don't want to introduce a bug so shortly before release (is it possible that there incomplete system translations for example?). 

> Only the en-GB translation in the translation folder used an underscore instead of a dash like the others. Does this work as well?  
> In the /usr/share/qt5/translations folder the country postfix is always separated by _

I think so. Just try with one of these:

    $ LANG=de_DE ./mixxx
    $ ./mixxx --locale de_DE

I didn't notice any issues when testing this.

uklotzde commented:
> Why do we tray to load the qt translations from the res path when we already have loaded them from the system? Can we make the local lookup conditionally? Or load directly from the correct location target depending.
> 
> On windows we install also qtbase and qthelp translations among others in the translations folder but the corresponding install statement is missing. I am afraid they are not loaded, are they? Are they loaded under Windows.
> 
> Only the en-GB translation in the translation folder used an underscore instead of a dash like the others. Does this work as well?
> In the /usr/share/qt5/translations folder the country postfix is always separated by _

https://doc.qt.io/qt-5/qlocale.html#QLocale-1

> The separator can be either underscore or a minus sign.

daschuer commented:
I am afraid really have a regression here. I am sorry that this PR is such hard.  

I have selected "Spanish (Mexico)" from the interface preferences in Mixxx 2.3. After restart, the Auto DJ tree knot shows the correct "DJ Automatico" from the mixxx_es-MX.ts file. After merging this branch, the Auto DJ knot is renamed to "Auto DJ" which is part of the mixxx_es.ts

By using and intermediate QLocale, Qt turns the - to _ and the translation is no longer found.

Holzhaus commented:
You're right, but I don't think this is a regression. Before, it also didn't work properly because it was required to use the correct separator (underscore vs. dash) to get the correct locale. That's why I didn't notice. And for System language, this also didn't work I guess.

There is no way to use the dash as delimiter without building the file name yourself:
https://doc.qt.io/qt-5/qtranslator.html#load-1

Instead, we should use the standard underscore for our translations IMHO. If you agree I'll rename the translation files, but we also need to figure out to configure tx correctly.

daschuer commented:
Renaming the files, works for me as a 2.3.0 band aid.
The loading of the additional qt translation can be postponed.

Holzhaus committed:
Translations: Use POSIX-style locale identifiers

Switch to using an underscore to separate ISO-639-1 language and
ISO-3166-1 country codes (instead of a dash).

Also removes the `lang_map` setting from the Transifex client
configuration that replaces the underscore with a dash:
https://docs.transifex.com/client/client-configuration#language-mappings

This fixes the broken loading of translations that used a dash in the
filename, because this is not supported by `QTranslator::load()`.
See https://doc.qt.io/qt-5/qtranslator.html#load-1 for details.

Holzhaus commented:
@daschuer Please check again, should be fixed now.

daschuer commented on .tx/config:
```diff
@@ -1,6 +1,5 @@
 [main]
 host = https://www.transifex.com
-lang_map = af_ZA: af-ZA, am_ET: am-ET, ar_AA: ar-AA, ar_AE: ar-AE, ar_BH: ar-BH, ar_DZ: ar-DZ, ar_EG: ar-EG, ar_IQ: ar-IQ, ar_JO: ar-JO, ar_KW: ar-KW, ar_LB: ar-LB, ar_LY: ar-LY, ar_MA: ar-MA, ar_OM: ar-OM, ar_QA: ar-QA, ar_SA: ar-SA, ar_SY: sy-IQ, ar_TN: ar-TN, ar_YE: ar-YE, arn_CL: arn-CL, as_IN: as-IN, ast_ES: ast-ES, az_AZ: az-AZ, ba_RU: ba-RU, be_BY: be-BY, bg_BG: bg-BG, bn_BD: bn-BD, bn_IN: ...
```
Thank you for digging down to the root cause.

daschuer commented:
I can confirm it is working now.  
As a side effect the selection falls back to system, because the dash version in the preferences is no longer found. 
But IMHO we can effort this, because the old state was kind of broken anyway. 

Is this issue worth a changelog entry?
